internal/resource/url.go

```diff
@@ -375,6 +388,36 @@ type s3target interface {
 	io.ReadSeeker
 }
 
+// getBucketRegion determines the partition and region a given S3 bucket is located in
+func (f *Fetcher) getBucketRegion(ctx context.Context, sess client.ConfigProvider, u url.URL) (string, error) {
+	// GetBucketRegion will return NotFound if the bucket doesn't exist in the same
+	// partition as the regionHint despite the default resolver in the session object
+	// supporting all partitions. First attempt to fetch from the S3RegionHint value
+	// on the Fetcher object if it exists, if not fallthrough to try with one region
+	// from each partition.
```

Comment:
It probably makes sense to keep a cache of regions for buckets we've seen.

Reply:
Hmm, what are you thinking for this, pre-pending to the list of buckets we search through so the order would turn into something like:

1. If being ran on AWS we first attempt with the region the server is located in
1. cached regions we've already consumed objects from
1. us-east-1 (US partition)
1. us-gov-east-1 (Govcloud)
1. cn-north-1 (aws-cn)

If so, what would the main goal be? To lower the loops required for each bucket if it's in a non-server + non-us partition? Or is there some other benefit for having the `regionHint` being dead-on (I'd assume the `GetBucketRegion` call itself might be a bit faster but I haven't tested if it makes any notable change in performance)?

Reply:
The cache would be a map from bucket name to region, and I'd put the cache lookup before anything else in this function.  The assumption is that a config will often have a lot of objects coming from the same bucket, and in that case we can avoid making any extra requests.

Reply:
Fair enough; my initial thinking was that in the worst case scenario (the bucket you're fetching from is in aws-cn and your server is not) you're doing 4 fetches with the current setup.

If we add every unique region we've encountered an object from to the list we can end up in scenarios where if you have lots of objects in different regions & partitions causing an excessive amount of attempts.

The extreme scenario would be that I have one object in every single region of every partition, the config specifies them in the order of one object from each us partition region, one object from each govcloud region, one object in aws-cn; when we attempt to fetch the object in aws-cn we are looking at X + Y + 4 fetches (where X is the amount of regions in the us partition & Y is the amount of regions in govcloud).